### PR TITLE
Revised is keyword

### LINKED ISSUE #1377: C# Language Reference keywords is
In C# 7, this now supports assignment in pattern matching expressions.

The migrated content does not include this usage.

### PR DESCRIPTION
**Pull request by rpetrusha**
# Revised `is` keyword

## Summary

Added information on support for pattern matching, and added additional detail on the "classic" `is` keyword

Fixes #1377

## Suggested Reviewers
@BillWagner

### CODE REVIEW

File: docs/csharp/language-reference/keywords/is.md [hunk 1]
```diff
@@ -31,35 +32,135 @@ translation.priority.ht:
...
+
+[!code-cs[is#1](../../../../samples/snippets/csharp/language-reference/keywords/is/is1.cs)]
+
+The `is` statement is true if:
+
+- *expr* is an instance of the same type as *type*.
+
+- *expr* is an instance of a type that derives from *type*. In other words, the result of *expr* can be upcast to an instance of *type*.
+
+- *expr* is an instance of a type that is a base class of *type*, and *expr* is the result of a downcast from *type* or one of its derived types.
```
Comment (BillWagner) [inline docs/csharp/language-reference/keywords/is.md]: This line reads wrong to me. I interpreted a few times that the `is` operator would return `true` when the object was a base class of the type being tested. (Example 3 shows that interpretation is wrong). How about:

- *expr* has a compile time type that is a base class of *type*, and *expr* has a runtime type that is *type* or is derived from *type*.

We may need to define *compile time type* and *runtime type* or point to the topic on the type system for that to stand on its own.

File: docs/csharp/language-reference/keywords/is.md [hunk 2]
```diff
@@ -31,35 +32,135 @@ translation.priority.ht:
...
+
+where *expr* is an expression that evaluates to an instance of some type, *type* is the name of the type to which the result of *expr* is to be converted, and *varname* is the object to which the result of *expr* is converted if the `is` test is `true`. 
+
+The `is` expression is `true` if any of the following is true:
+
+- *expr* is an instance of the same type as *type*.
+
+- *expr* is an instance of a type that derives from *type*. In other words, the result of *expr* can be upcast to an instance of *type*.
+
+- *expr* is an instance of a type that is a base class of *type*, and *expr* is the result of a downcast from *type* or one of its derived types.
```
Comment (BillWagner) [inline docs/csharp/language-reference/keywords/is.md]: Same comment as above (for a similar sentence.)

File: docs/csharp/language-reference/keywords/is.md [hunk 3]
```diff
@@ -31,35 +32,135 @@ translation.priority.ht:
...
+
+- *expr* is an instance of a type that implements the *type* interface.
+
+If *exp* is `true` and `is` is used with an `if` statement, *varname* is assigned and has local scope within the `if` statement only.
+
+The following example uses the `is` type pattern to provide the implementation of a type's <xref:System.IComparable.CompareTo(System.Object)?displayProperty=fullName> method.
+
+[!code-cs[is#5](../../../../samples/snippets/csharp/language-reference/keywords/is/is-type-pattern5.cs)]
+
+Without pattern matching, this code might be written as follows. The use of type pattern matching produces more compact, readable code by eliminating the need to test whether the result of a conversion is a `null`.
```
Comment (BillWagner) [inline docs/csharp/language-reference/keywords/is.md]: I'd consider adding an example using is with a value type, because the previous version of that could would require `as` and an explicit cast, or a cast inside a `try` / `catch` block.

**Review by BillWagner — CHANGES_REQUESTED** [review]
@rpetrusha Overall, this looks great.

There was one sentence that I mis-interpreted upon reading. I suggested an alternative. Or, if you come up with a better way, that's fine too.
Also, a couple suggestions to consider or leave out.

File: docs/csharp/language-reference/keywords/is.md (hunk 1 above)
Comment (rpetrusha) [inline docs/csharp/language-reference/keywords/is.md]: @BillWagner I've revised it. Incredibly, I couldn't find a definition of compile-time type and runtime type anywhere in our doc set (other than in a blog). I think I'll open an issue so that, at some point, its absence can be addressed in the docs.

File: docs/csharp/language-reference/keywords/is.md (hunk 3 above)
Comment (rpetrusha) [inline docs/csharp/language-reference/keywords/is.md]: @BillWagner I've added an example.

File: docs/csharp/language-reference/keywords/is.md [hunk 4]
```diff
@@ -31,35 +32,142 @@ translation.priority.ht:
...
+
+[!code-cs[is#1](../../../../samples/snippets/csharp/language-reference/keywords/is/is1.cs#1)]
+
+The `is` statement is true if:
+
+- *expr* is an instance of the same type as *type*.
+
+- *expr* is an instance of a type that derives from *type*. In other words, the result of *expr* can be upcast to an instance of *type*.
+
+- *expr* has a compile-time type that is a base class of *type*, and *expr* has a runtime type that is *type* or is derived from *type*. The *compile-time type* of a variable is the variable's type as defined in its type declaration. The *runtime type* of a variable is the type of the instance that is assigned to that variable.
```
Comment (BillWagner) [inline docs/csharp/language-reference/keywords/is.md]: You should remove 'type' from "in its type declaration". The compile time type of a variable is its type as defined in its declaration.

File: docs/csharp/language-reference/keywords/is.md [hunk 5]
```diff
@@ -31,35 +32,142 @@ translation.priority.ht:
...
+
+where *expr* is an expression that evaluates to an instance of some type, *type* is the name of the type to which the result of *expr* is to be converted, and *varname* is the object to which the result of *expr* is converted if the `is` test is `true`. 
+
+The `is` expression is `true` if any of the following is true:
+
+- *expr* is an instance of the same type as *type*.
+
+- *expr* is an instance of a type that derives from *type*. In other words, the result of *expr* can be upcast to an instance of *type*.
+
+- *expr* has a compile-time type that is a base class of *type*, and *expr* has a runtime type that is *type* or is derived from *type*. The *compile-time type* of a variable is the variable's type as defined in its type declaration. The *runtime type* of a variable is the type of the instance that is assigned to that variable.
```
Comment (BillWagner) [inline docs/csharp/language-reference/keywords/is.md]: Same nit as above.

**Review by BillWagner — APPROVED** [review]

**Comment by BillWagner** [comment]
@rpetrusha The new version looks great. There's one nit I left to fix in the changes. After that, let's :shipit:
